Fix route helper to work with string params

codeodor opened the pull request:
This is a redo of #18173, with the change made against the master branch, as requested by @sgrif.

Before Rails 4.2.0, `model_path("id" => 1)` would work fine. As of Rails 4.2.0, it generates a couple of errors:
1. `ArgumentError: comparison of Array with Array failed` (due to attempting to sort a hash with mixed string/symbol keys)
2. `ActionController::UrlGenerationError`, because of missing required key :id (i.e., it ignored the "id" string parameter)

codeodor committed:
Fix route helper to work with string params

Before Rails 4.2.0, `model_path("id" => 1)` would work fine. As of Rails 4.2.0, it generates a couple of errors:

1. ArgumentError: comparison of Array with Array failed (due to attempting to sort a hash with mixed string/symbol keys)
2. ActionController::UrlGenerationError, because of missing required key :id (i.e., it ignored the "id" string parameter)

This fix ensures all string keys to route helpers are converted to symbols and deprecated with an appropriate message.

rafaelfranca commented:
Could we remove the code that deprecate `'action'` and `'controller'` after this change?

codeodor commented:
@rafaelfranca I'm assuming that's a question for someone else participating in this ticket, but just in case, I thought I'd chime in to let you know I have no idea. :smile:

rafaelfranca commented:
Actually it was for you. This code deprecates string option to url_for helpers. https://github.com/rails/rails/blob/9598c9655f04b3bf6dfc245437ad5a09f23c792c/actionpack/lib/action_dispatch/routing/route_set.rb#L305-317

If we are deprecating all string options I guess we can remove that code or at least move your implementation to that method.

codeodor commented:
Interesting, thanks. I didn't see it before. Let me see how they can work together, because clearly that method wasn't being called before. I will update again when I've had a chance to look in detail.

codeodor commented:
The answer was yes, we could remove it. However, it seemed as good a place as any to put the deprecation notice, so I just had it apply to all keys, and removed my deprecation notice, which still fixes the issue.

On the other hand, since you pointed me to it I started looking into that commit, and I found it arose to fix https://github.com/rails/rails/issues/16958, where there was some discussion on the security risk of converting an unknown number of strings to symbols.

I don't know if it applies though, since presumably the same number of symbols were already there from the request (if we simply passed `params` to the function, for instance). Since this fix is only needed when we pass strings into the helper, is that a worry here? Or can we even be sure that `params` has symbol keys here?

If it is a worry, should we show the deprecation notice and convert to strings instead? Then we could change to symbols in Rails 5 when Ruby 2.2 is required?

Sam

codeodor commented:
@rafaelfranca any thoughts?

sgrif commented:
@schneems @rafaelfranca Any objections to merging this?

schneems commented:
On a technical level i'm fine with the code. I'm not sure about the policy level of releasing deprecations, i'm also still a little nonplussed about speed implications here. If the plan is to only keep it in for a short time and then to switch to only symbols i'm :+1:

sgrif commented:
I think this might need some additional discussion actually, as I recall there being a conversation about whether we jumped the gun on switching to symbol keys in campfire the other day, due to passing `params.merge(whatever)` being a common thing to do.

codeodor commented:
@sgrif If the discussion was about jumping the gun on switching to symbol keys, I think this PR is the fix for that. Perhaps you all talked about changing all code to use string keys instead, in which case this one won't help, but it does fix 2 bugs and allows for both styles of keys.

codeodor commented:
Since this had been sitting for a while and is marked as a release blocker, I went ahead and rebased it against master so you can merge more easily.

codeodor commented:
(I also ran the test against master to be sure it was still relevant -- it failed and passes with the fix).

codeodor commented:
Ok, well apparently I need to double check that to see why the test failed.

codeodor commented:
So actually @y00rb made this change last week: https://github.com/rails/rails/commit/a77de09812eec5f342dd159f053038c6f67e6da4

And this PR causes that test to fail.

I will remove that test, because we determined in the discussion above that behavior is a breaking change that should have a deprecation notice.

However, I will leave it in a 2nd commit rather than rebasing into a single one, that way the history is preserved.

codeodor committed:
Remove test that mixed keys should raise error in routing
The test is testing behavior that breaks documented functionality that has yet to be deprecated. See #18175 for discussion.

codeodor commented:
I'll wait to hear back before fixing the merge conflicts. I have a feeling it's just the change log, because it changes so frequently. :smile:

y00rb commented:
Is it cause by my test?

codeodor commented:
No, your test is fine and asserts what the behavior is supposed to be in Rails 5.0 (as I understand it).

But for purposes of this pull request and Rails before it reaches 5.0, the behavior your test verifies actually breaks something that has not yet been deprecated. So this pull request deprecates that behavior, and should not be applied to 5.0 -- only 4.2.

y00rb commented:
got it, so will deprecates string option to url_for helpers in 5.0, right?
if deprecates string option, I think it is avoid mixed & replicated parameters which I fixed problem.

codeodor commented:
I think the original plan was that this would have been deprecated already, then removed in 5.0. But this spot was missed, so there was talk that it would be OK to deprecate it in 4.2.1 even though it's late for a deprecation notice. But 4.2.1 is out, so maybe 4.2.2?

y00rb commented:
gotcha

agis- commented:
So what's the current state of this? Isn't this already fixed in https://github.com/rails/rails/commit/4d84922840deb89754f5b5fb61ebea0aeadc52df?

I see this is still an issue however when I've upgraded from 4.1 to 4.2, in various places where we were doing `some_path(params.merge(foo: "bar"))` and the params contain string keys.

So what's the way to go now? Should we just call `#symbolize_keys` on the params or?

cc @rafaelfranca @codeodor @sgrif

y00rb commented:
@agis- 
You could upgrade rails to 4.2.1 or 4.2.2, the change doesn't include 4.2

agis- commented:
@y00rb I'm on 4.2.4.

codeodor commented:
@agis-  that commit did the deprecation but it was not complete, so as you see this commit is still waiting to finish off the job and your string keys will fail until this, or if another fix exists, is merged.

I just monkey patched my apps for now, or as you say you can convert all your hashes to use symbol keys in these cases. I would prefer to see this fixed, but for now that's what I've done.

agis- commented:
@pixeltrix is there any chance that we see this merged anytime soon? Are there other things that need to be done? /cc @rafaelfranca

maclover7 commented:
IMHO, I don't think we should be accepting this PR (targeting the master branch), due to that string keys are already deprecated via https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/routing/route_set.rb#L310-L325 on 4-2-stable, and the deprecation notice has been dutifully removed in time for 5.0 on master. Merging this PR would cause the deprecated behavior to be reintroduced on the master (5.0) branch. If there is a bug as far as how the deprecation is managed on 4-2-stable, then I'd say to open up a PR against 4-2-stable.
